app/controllers/incomeSources/cease/CheckCeaseForeignPropertyDetailsController.scala

```diff
@@ -36,28 +41,137 @@ class CheckCeaseForeignPropertyDetailsController @Inject()(val authenticate: Aut
   def showAgent(): Action[AnyContent] = Authenticated.async {
     implicit request =>
       implicit user =>
         getMtdItUserWithIncomeSources(incomeSourceDetailsService).flatMap {
           implicit mtdItUser =>
-            Future.successful(NotImplemented)
+            handleRequest(
+              isAgent = true,
+              backUrl = backUrlAgent,
+              changeUrl = changeUrlAgent,
+              homePageCall = homePageCallAgent,
+              itvcErrorHandler = itvcErrorHandlerAgent
+            )
+        }
+  }
+
+  def submit(cessationDate: String): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit request =>
+      handleSubmitRequest(
+        successCall = successCall,
+        homePageCall = homePageCall,
+        cessationDate = cessationDate,
+        itvcErrorHandler = itvcErrorHandler
+      )
+  }
+
+  def submitAgent(cessationDate: String): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleSubmitRequest(
+              cessationDate = cessationDate,
+              successCall = successCallAgent,
+              homePageCall = homePageCallAgent,
+              itvcErrorHandler = itvcErrorHandlerAgent
+            )
         }
   }
 
+  def handleRequest(isAgent: Boolean,
+                    backUrl: String,
+                    changeUrl: String,
+                    homePageCall: Call,
+                    itvcErrorHandler: ShowInternalServerError)
+                   (implicit user: MtdItUser[_], request: Request[_]): Future[Result] = {
+
+    if (isDisabled(IncomeSources)) {
+      Future.successful(Redirect(homePageCall))
+    } else {
+      request.session.get(ceaseForeignPropertyEndDate) match {
+        case Some(date) =>
+          Future(Ok(checkCeaseForeignPropertyDetails(
+            endDate = date,
+            backUrl = backUrl,
+            isAgent = isAgent,
+            changeUrl = changeUrl,
+            formattedEndDate = longDate(date.toLocalDate).toLongDate
+          )))
+        case _ =>
+          Logger("application").error(s"[CheckCeaseForeignPropertyDetailsController][handleSubmitRequest]:" +
+            s" Could not get ceaseForeignPropertyEndDate from session")
+          Future(itvcErrorHandler.showInternalServerError())
+      }
+    }
+  }
+
+  def handleSubmitRequest(homePageCall: Call,
+                          successCall: Call,
+                          cessationDate: String,
+                          itvcErrorHandler: ShowInternalServerError)
+                         (implicit user: MtdItUser[_]): Future[Result] = {
+
+    if (isDisabled(IncomeSources)) {
+      Future.successful(Redirect(homePageCall))
+    } else {
+
+      val maybeIncomeSourceId = user.incomeSources.properties
```

Comment:
- I would assume this is some sort of logic to verify that this incomeSource exists before executing Runtime logic to drop incomeSource?  
- Because we intended incomeSource to be cached in the user object, this part might not work well (caching might not be up to date etc);  this might be okis to fully reply on the call to drop incomeSource without checking if this incomeSource exist;